federation/pkg/kubefed/init/init.go

```diff
@@ -245,24 +246,24 @@ func (i *initFederation) Run(cmdOut io.Writer, config util.AdminConfig) error {
 	// 2. Expose a network endpoint for the federation API server
 	svc, ips, hostnames, err := createService(hostClientset, i.commonOptions.FederationSystemNamespace, serverName, i.options.apiServerAdvertiseAddress, i.options.apiServerServiceType, i.options.dryRun)
 	if err != nil {
-		return err
+		return cleanupObjects(hostClientset, i.commonOptions.FederationSystemNamespace, i.options.dryRun, err)
```

Comment:
I think having to call  ``cleanupObjects`` is a violation of DRY.  Consider factoring the contents of Run into a new method and have Run call that method.  If an error is returned by the method, perform cleanup.